---
title: "Single cell perspective into Extracellular Vesicules projected onto Spatial Transcriptomics"
format:
  html:
    code-fold: true
    code-tools: true
    embed-resources: true
    self-contained-math: true
jupyter: python3
---

# Load in data

Set up the environment and load the deps:

```{python}
#| output: false

import re
import os
import glob
import warnings
import pandas as pd
import numpy as np
import scanpy as sc
import squidpy as sq
import anndata as an
from pathlib import Path
from dotenv import load_dotenv
from scipy.sparse import csr_matrix
import matplotlib.pyplot as plt
import seaborn as sns

# pd.set_option('display.max_columns', 100)

## import the functions used in this report (got too long to include them in the report)
try: from analysis.helpers.projection_functions import *
except ImportError: from helpers.projection_functions import *

## load the environment variables from the .env file
load_dotenv()
if os.getenv("DATA_FOLDER") is None:
  load_dotenv(Path.cwd() / ".env")

anndata_folder = os.getenv("ANNDATA_FOLDER")
atlas_folder = os.getenv("ATLAS_FOLDER")
data_folder = os.getenv("DATA_FOLDER")
st_folder = os.getenv("ST_FOLDER")
model_folder = Path(data_folder).expanduser() / "saved_models"
raw_data_folder = os.getenv("RAW_DATA_FOLDER")
snapshot_an_loc = Path(data_folder).expanduser() / "sc_deconv_snapshot.h5ad"

```


Load in the spatial slides experiments

```{python}
# Load the data
data_directory = Path(anndata_folder).expanduser() # Current directory, adjust as needed
adata_collection = load_spatial_data(data_directory)

phases = {
  "Proliferative": ["152806", "152810"],
  "Early-Secretory": ["152807", "152811"],
  "Mid-Secretory": ["CTR1_processed_data", "CTR2_processed_data", "CTR3_processed_data", "CTR4_processed_data"]
  }
```


# Look at specific gene expression profiles in spatial transcriptomic slides

## Fibroblasts in myometrium

From Elina:

>Ma oletan, et äkki näeme strooma rakke projitseerituna müomeetriumi, sest seal võivad reaalselt olla fibroblastid.
>
>Leidsin selle artikli: https://academic.oup.com/molehr/article/20/3/250/1114116
>
>Kas sa saaksid vaadata, kas need "strooma rakud" mis asuvad müomeetriumis ekspresseerivad geene CD90, ALDH1A1, ALDH1A2, ALDH1A3, ALDH1B1?

First let's project the abundances of the stromal cell types on the slide with the scRNA-seq data. We are using slide with id 152807 (Early-Secretory)

::: {.panel-tabset}
### Reference
```{python}
vis_dat = adata_collection[("152807", "ref")].copy()
sq.pl.spatial_scatter(vis_dat, cmap='magma', ncols=2,
  color=vis_dat.obs.columns[vis_dat.obs.columns.str.contains("stroma", case=False)])
```
### EV
```{python}
vis_dat = adata_collection[("152807", "ev")].copy()
sq.pl.spatial_scatter(vis_dat, cmap='magma', ncols=2,
  color=vis_dat.obs.columns[vis_dat.obs.columns.str.contains("stroma", case=False)])
```
### Biopsy
```{python}
vis_dat = adata_collection[("152807", "bio")].copy()
sq.pl.spatial_scatter(vis_dat, cmap='magma', ncols=2,
  color=vis_dat.obs.columns[vis_dat.obs.columns.str.contains("stroma", case=False)])
```
:::

Then we look at specific gene expression profiles in the slide, the genes were interested in are CD90, ALDH1A1, ALDH1A2, ALDH1A3, ALDH1B1. CD90 does not seem to be in the dataset, mightve been filtered out via the highly variable gene filtering or it has an alternative name here. We do have an aliased genes for the CD90, the THY1.

For this view, we do not to look at different projections, as we are looking at the expression of the genes in the original spatial transcrpitomic slide itself.

```{python}
slide_id = "152807"
vis_dat = adata_collection[("152807", "ref")].copy()
genes_of_interest = ["THY1", "ALDH1A1", "ALDH1A2", "ALDH1A3", "ALDH1B1"]
# the original unfiltered slide is stored within the ".raw" object
sq.pl.spatial_scatter(vis_dat, use_raw=True, cmap='magma', ncols=2, color=genes_of_interest)
```

## EV biomarker gene projection

In addition, lets look at some EV biomarker genes as well. As we have a lot of slides, we are looking at only a selection of them, ie one per cycle phase to project, and plot the gene group scores.

```{python}
## Create a dictionary of all the genes we want to plot, and then go per key
all_genes = {
  "ESCRT-0": ["HGS", "STAM", "STAM2"],
  "ESCRT-I": ["MVB12A", "MVB12B", "TSG101", "UBAP1", "VPS28", "VPS37A", "VPS37B", "VPS37C", "VPS37D"],
  "ESCRT-II": ["SNF8", "VPS25", "VPS36"],
  "ESCRT-III": ["CHMP2A", "CHMP2B", "CHMP3", "CHMP4A", "CHMP4B", "CHMP4C", "CHMP6"],
  "ESCRT-IV": ["VPS4A","VPS4B","VTA1"],
  "ESCRT-independent": ["RAB31", "SMPD3", "ALIX"],
  "EV release gene markers": ["RAB27A", "RAB27B", "RAB11A", "RAB11B", "RAB3A", "RAB3B", "RAB3C", "RAB3D", "RAB7A", "RAB7B"]
}

print("Gene groups:", all_genes)

```

::: {.panel-tabset}

### Proliferative

```{python}
# select the slide
vis_dat = adata_collection[("152806", "ref")].copy()
vis_dat = calculate_scores(vis_dat, all_genes)

sq.pl.spatial_scatter(
  vis_dat, use_raw=True, cmap='magma', ncols=2, color=all_genes.keys()
  )
```

### Early-secretory

```{python}
# select the slide
vis_dat = adata_collection[("152807", "ref")].copy()
vis_dat = calculate_scores(vis_dat, all_genes)

sq.pl.spatial_scatter(
  vis_dat, use_raw=True, cmap='magma', ncols=2, color=all_genes.keys()
  )
```

### Mid-secretory

```{python}
# select the slide
vis_dat = adata_collection[("CTR4_processed_data", "ref")].copy()
vis_dat = calculate_scores(vis_dat, all_genes)

sq.pl.spatial_scatter(
  vis_dat, use_raw=True, cmap='magma', ncols=2, color=all_genes.keys()
  )
```

:::

## Correlation between cell profile and EV biomarkers

Here we are interested in the correlation between the expression of EV biomarkers and the cell profiles in the spatial transcriptomic data. This can help us understand how different cell types contribute to EV production and release in the tissue context.

Here, we calculate correlation with the gene group score from previous chapter, and also look at more finetuned gene-specific correlation. For basic approach, we use score_genes function from scanpy to summarise groups of genes expression into more digestible format.

Here we show an example of a early secretory slide, id "152807", with various statistical tests

```{python}
# 1. Set values
OBSM_KEY = "tangram_ct_pred"
slide_ids = np.unique([slide[0] for slide in adata_collection.keys()])
# print(slide_ids)
slide_id = slide_ids[1]
vis_dat = adata_collection[(slide_id, "ref")].copy()
vis_dat = calculate_scores(vis_dat, all_genes)

# 2. Calculate Correlations
gene_groups = list(all_genes.keys())
pearson_mat, jaccard_mat, moran_mat, moran_pval_mat = calculate_correlations(
  vis_dat, 
  gene_groups=gene_groups, 
  obsm_key=OBSM_KEY,
  n_perms=1000 
)

# 3. Plot Heatmaps
plot_correlation_heatmap(pearson_mat, f"Slide {slide_id}", "Pearson Correlation")
plot_correlation_heatmap(jaccard_mat, f"Slide {slide_id}", "Jaccard Index (>75%)")
# plot_correlation_heatmap(moran_mat, f"Slide {slide_id}", "Bivariate Moran's I")

# 3. Bivariate Moran's I (WITH Masking)
# Pass the p-value matrix computed in the previous step
plot_correlation_heatmap(
    moran_mat, 
    f"Slide {slide_id}", 
    "Bivariate Moran's I (Significant Only)", 
    pval_df=moran_pval_mat, 
    sig_threshold=0.05
)

```


::: {.panel-tabset}

### Proliferative

```{python}

# 1. Set values
OBSM_KEY = "tangram_ct_pred"

```

```{python}
slide_ids = phases["Proliferative"]
# Run Aggregation
agg_pearson, agg_jaccard, agg_moran, agg_pval = aggregate_correlations(
  slide_ids,
  adata_collection, # Your main dictionary of slides
  all_genes,        # Your gene dictionary
  OBSM_KEY
)
results_prolif = (agg_moran, agg_pval)

# Plot the Aggregated Results
plot_correlation_heatmap(
    agg_moran, 
    "Proliferative Phase (Combined)", 
    "Avg Bivariate Moran's I", 
    pval_df=agg_pval, 
    sig_threshold=0.05
)
```

### Early-secretory

```{python}
slide_ids = phases["Early-Secretory"]

# Run Aggregation
agg_pearson, agg_jaccard, agg_moran, agg_pval = aggregate_correlations(
  slide_ids,
  adata_collection, # Your main dictionary of slides
  all_genes,        # Your gene dictionary
  OBSM_KEY
)
results_early = (agg_moran, agg_pval)

# Plot the Aggregated Results
plot_correlation_heatmap(
    agg_moran, 
    "Early Secretory Phase (Combined)", 
    "Avg Bivariate Moran's I", 
    pval_df=agg_pval, 
    sig_threshold=0.05
)
```

### Mid-secretory

```{python}
slide_ids = phases["Mid-Secretory"]

# Run Aggregation
agg_pearson, agg_jaccard, agg_moran, agg_pval = aggregate_correlations(
  slide_ids,
  adata_collection, # Your main dictionary of slides
  all_genes,        # Your gene dictionary
  OBSM_KEY
)
results_mid = (agg_moran, agg_pval)

# Plot the Aggregated Results
plot_correlation_heatmap(
  agg_moran, 
  "Mid Secretory Phase (Combined)",
  "Avg Bivariate Moran's I",
  pval_df=agg_pval,
  sig_threshold=0.05
)
```

:::

## Summing up the results

Here we take the Morans I results and summarize them in a more digestable visual format.

Lets make a line plot over cycle phases, take only differentially significant correlations into account, and plot a line plot for each gene group. 

```{python}
## Load the single cell reference data lineage information to group cell types
sc_dat = an.read_h5ad(snapshot_an_loc, backed="r")
sc_obs = sc_dat.obs.copy()
del sc_dat
sc_obs.replace({"lineage": {"Mesenchymal": "Stromal"}}, inplace=True)
general_lineages = sc_obs[["celltype", "lineage"]].drop_duplicates().set_index("celltype")

# Organize results in order
phase_results = {
    'Proliferative': results_prolif, 
    'Early-Secretory': results_early, 
    'Mid-Secretory': results_mid
}
```

Lets divide this summarized view into different cell lineages.

::: {.panel-tabset}

### Stromal

```{python}

stromal_cell_types = general_lineages[general_lineages['lineage'] == 'Stromal'].index.tolist()
plot_nested_dot_plot(phase_results, cell_type_order=stromal_cell_types, sig_threshold=0.05)

```

### Epithelial

```{python}

epi_cell_types = general_lineages[general_lineages['lineage'] == 'Epithelial'].index.tolist()
plot_nested_dot_plot(phase_results, cell_type_order=epi_cell_types, sig_threshold=0.05)

```

### Immune & Endothelial

Put together as they have less cell types, and we want to be able to compare them more easily.

```{python}

immune_cell_types = general_lineages.query("lineage == 'Immune' or lineage == 'Endothelial'").index.tolist()
plot_nested_dot_plot(phase_results, cell_type_order=immune_cell_types, sig_threshold=0.05)

```

Maybe not include this in the final plot, the frame has wrong size here. Additinoally, these cell types are not the main focus here, and Visium experiment wouldnt cover them well anyway.

:::

## Using an unified scoring for all the genes

Here we use the `scanpy.tl.score_genes` function to summarise all the EV-related genes together, and focus on general patterns throughout the cycle phases.

```{python}

unified_genes = {"EV-combined": [item for sublist in all_genes.values() for item in sublist]}
cycle_all_ev = {}
for phase in phases.keys():
  _, _, agg_moran, agg_pval = aggregate_correlations(phases[phase], adata_collection, unified_genes, OBSM_KEY)
  phase_stats = (agg_moran, agg_pval)
  cycle_all_ev[phase] = phase_stats

plot_ev_phase_heatmap(
    cycle_all_ev, 
    gene_col="EV-combined", 
#     cell_type_list=general_lineages.index.tolist(), 
    lineage_info=general_lineages, 
    fdr_threshold=0.05
)

```