---
title: "Single cell perspective into Extracellular Vesicules projected onto Spatial Transcriptomics"
format:
  html:
    code-fold: true
    code-tools: true
    embed-resources: true
    self-contained-math: true
jupyter: python3
---

# Load in data

Set up the environment and load the deps:

```{python}
#| output: false

import re
import pandas as pd
pd.set_option('display.max_columns', 100)
import scanpy as sc
import squidpy as sq
import anndata as an
import numpy as np
import seaborn as sns
import seaborn.objects as so
import matplotlib.pyplot as plt
from pathlib import Path
from os import getenv
from dotenv import load_dotenv
from scipy.sparse import csr_matrix
from scipy.stats import ttest_ind
from scipy.stats import spearmanr
from scipy.stats import f_oneway
from scipy.stats import tukey_hsd
from statsmodels.stats.multicomp import pairwise_tukeyhsd
from sklearn.decomposition import PCA
from scipy.cluster.hierarchy import linkage, dendrogram, fcluster
from matplotlib.gridspec import GridSpec
from matplotlib.patches import Patch


## load the environment variables from the .env file
load_dotenv()
if getenv("DATA_FOLDER") is None:
  load_dotenv(Path.cwd() / ".env")

anndata_folder = getenv("ANNDATA_FOLDER")
atlas_folder = getenv("ATLAS_FOLDER")
data_folder = getenv("DATA_FOLDER")
st_folder = getenv("ST_FOLDER")
model_folder = Path(data_folder).expanduser() / "saved_models"
raw_data_folder = getenv("RAW_DATA_FOLDER")

```

Load in the data, filter out "Hormones" celltypes and non-endometriotic samples and finally do some preprocessing:

```{python}
#| output: false

## load in the modified HECA atlas
snapshot_an_loc = Path(data_folder).expanduser() / "sc_deconv_snapshot.h5ad"
sc_dat = an.read_h5ad(snapshot_an_loc)

## load in the EV CCHT only data and drop the "none" gene id in the end
ccht_uf_raw = pd.read_feather(Path(data_folder).expanduser() / "filtered" / "annot_raw.feather").set_index("gene_id").iloc[:-1]
ccht_uf_pheno = pd.read_table(Path(data_folder).expanduser() / "filtered" / "phenotype.tsv").set_index("samplename")
## Filter out some samples that are not behaving very well...
terminator = ["HUT26_UF", "HUT26_biopsy"]
ccht_uf_pheno = ccht_uf_pheno.query("samplename not in @terminator")
ccht_uf_raw = ccht_uf_raw.drop(columns=terminator)

```

Load in the spatial slides experiments

```{python}

slides_path = Path(st_folder).expanduser()
vis_dats = {}
for vis_fold in slides_path.iterdir():
  if vis_fold.is_dir():
    slide_id = vis_fold.stem
    vis_dat = an.read_h5ad(Path(anndata_folder) / f"st_slide_{slide_id}.h5ad")
    vis_dats[slide_id] = vis_dat

vis_keys = list(vis_dats.keys())

```

# Bulk deconvolution and projection onto single cells

We'd be interested in using scRNA-seq atlas to deconvolve the EV data. This helps us understand where are those EV produced and which tissue signal they are capturing. To run single cell and projection, we are gonna use [TAPE](https://www.nature.com/articles/s41467-022-34550-9) algorithm. This algorithm also enables us to use the trained VAE model to estimate the number and the distribution of original cell populations.

First we'll need to reformat the raw bulk RNA samples from EV to matrix form and prepare some plotting functions.

Train the model on the single cell reference data and generate the pseudo-single-cell data from the bulk samples. Here we divide the original dataset into "biopsy" and "EV" groups, which then summarises all of the corresponding sampling method samples into one representative group. This would also increase the confidence of cell profiles while normalising individual variance.

::: {.panel-tabset}

## EV group

```{python}

gen_sc_file = Path(anndata_folder).expanduser() / "gen_sc_ev.h5ad"
gen_ccht_ev = sc.read_h5ad(gen_sc_file, backed="r").to_memory()

```

The generated dataset is then preprocessed, scaled, embedded into lower dimensions and clustered via Leiden. The visualisation is done using celltype annotation from HECA

```{python}

sc.pl.embedding(
  gen_ccht_ev,
  basis='X_pca',
  color='celltype',
  frameon='small'
)

sc.pl.embedding(
  gen_ccht_ev,
  basis='X_umap',
  color='celltype',
  frameon='small'
)

sc.pl.embedding(
  gen_ccht_ev,
  basis='X_umap',
  color='lineage',
  frameon='small'
)

```

## Biopsy group

```{python}

gen_sc_file = Path(anndata_folder).expanduser() / "gen_sc_biopsy.h5ad"
gen_ccht_bio = sc.read_h5ad(gen_sc_file, backed="r").to_memory()

```

The generated dataset is then preprocessed, scaled, embedded into lower dimensions and clustered via Leiden.

```{python}

sc.pl.embedding(
  gen_ccht_bio,
  basis='X_pca',
  color='celltype',
  frameon='small'
)

sc.pl.embedding(
  gen_ccht_bio,
  basis='X_umap',
  color='celltype',
  frameon='small'
)

sc.pl.embedding(
  gen_ccht_bio,
  basis='X_umap',
  color='lineage',
  frameon='small'
)

```

:::


# Projection of generated counts to Spatial Transcriptomic slides

And now use the generated pseudo-single-cell dataset and project it to the spatial transcriptomic slide.

First we will go with the early secretory stage samples

::: {.panel-tabset}

```{python}
# set the exemplary sample to use for downstream tasks
example_sample = "152807" # "E184-1_bottom"
```

{{< include ev2space_slide.qmd >}}

```{python}
# set the exemplary sample to use for downstream tasks
example_sample = "152811" # "E184-1_bottom"
```

{{< include ev2space_slide.qmd >}}

:::

Then let's also look at the proliferative stage samples

::: {.panel-tabset}

```{python}
# set the exemplary sample to use for downstream tasks
example_sample = "152806" # "E184-1_bottom"
```

{{< include ev2space_slide.qmd >}}

```{python}
# set the exemplary sample to use for downstream tasks
example_sample = "152810" # "E184-1_bottom"
```

{{< include ev2space_slide.qmd >}}

:::

# Calculate summary statistics for the ST slides

Here we try to quantify the difference in distances of different enrichment signatures. For this, we calculate the following statistics:

* Per cell type over all the slides
  * Simple absolute difference of enrichment per spot
  * Correlation between slides via `scipy.signal.correlate2d`
  * Apply 2D filtering and then calculate the average distance to enrichment cluster on another projection
    * Try either mean or max filtering for the most enriched cell type
* General clustering statistics over all the slide
  * Moran's I which indicates the strength of clustering, here applied on celltype abundances with the `squidpy.gr.spatial_autocorr` function
  * Ripley's K-function on the maximum enrichment cell type, calculated via `squidpy.gr.ripley`
  * Jaccard Index compared to the `ref` slide after filtering enrichment scores with a lower quartile threshold
  * Total distance (Wasserstein distance?) per celltype over the whole slide per celltype

For the filtering step we wish to remove any lowly enriched clusters so that only the more highly expressed features are left. This step is also applied before the multiclass plots so that the lowly enriched values wouldnt affect the other cell types. The lowly enriched region filtering is performed via a simple thresholding approach where the threshold is the lower quartile of the distribution over the slide for the cell type.

In addition, we have subset some of the more niche celltypes that either don't exhibit a patterns in expression or are not related to the current cycle phase.

For the following analysis, we will look at only the early secretory samples - 152807 and 152811.

```{python}

## read the datasets in again
gen_st_file = Path(anndata_folder).expanduser() / "c2l_152807_ev.h5ad"
vis_dat_ev = sc.read_h5ad(gen_st_file, backed="r").to_memory()
gen_st_file = Path(anndata_folder).expanduser() / "c2l_152807_bio.h5ad"
vis_dat_bio = sc.read_h5ad(gen_st_file, backed="r").to_memory()
gen_st_file = Path(anndata_folder).expanduser() / "c2l_152807_ref.h5ad"
vis_dat_ref = sc.read_h5ad(gen_st_file, backed="r").to_memory()

```


```{python}

sq.gr.spatial_neighbors(vis_dat_ev, spatial_key="spatial")
the_conditional = " \
  celltype not in @vis_dat_ev.var.index & \
  celltype in @vis_dat_ev.obs.columns \
"
# avoid some errors from celltypes having same name as genes
cts = gen_ccht_ev.obs.query(the_conditional).celltype.unique()

# filter out celltypes with good clustering
morans_autocorr = sq.gr.spatial_autocorr(
  vis_dat_ev,
  genes=cts,
  attr="obs",
  mode="geary",
  n_perms=1000,
  n_jobs=8,
  copy=True,
  transformation=True
)
# Filter for significant p-values and positive I scores (indicating clustering)
signf_ct = morans_autocorr.query("pval_sim_fdr_bh < 0.05 and C > 0").index

```

```{python}

sq.pl.spatial_scatter(
  vis_dat_ev,
  color=signf_ct,
  spatial_key="spatial",
  shape="hex",
  img=False
)

```

